pkg/mailboxwatch/impl.go

```diff
@@ -0,0 +1,200 @@
+/*
+Copyright 2023 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mailboxwatch
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	machmeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	k8ssets "k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/watch"
+	upstreamcache "k8s.io/client-go/tools/cache"
+
+	tenancyv1a1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	tenancyv1a1informers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/tenancy/v1alpha1"
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	"github.com/kcp-dev/edge-mc/pkg/placement"
+)
+
+func newFilteredListerWatcher(
+	mailboxWorkspacePreInformer tenancyv1a1informers.WorkspaceInformer,
+	unfiltered upstreamcache.ListerWatcher,
+) *filteredListerWatcher {
+	flw := &filteredListerWatcher{
+		unfiltered: unfiltered,
+		clusters:   k8ssets.NewString(),
+	}
+	mailboxWorkspacePreInformer.Informer().AddEventHandler(flw)
+	return flw
+}
+
+type filteredListerWatcher struct {
+	sync.Mutex
+	unfiltered upstreamcache.ListerWatcher
+	clusters   k8ssets.String
+}
+
+func (flw *filteredListerWatcher) OnAdd(obj any) {
+	flw.setInclusion(obj, true)
+}
+
+func (flw *filteredListerWatcher) OnUpdate(old, obj any) {
+	flw.setInclusion(obj, true)
+}
+
+func (flw *filteredListerWatcher) OnDelete(obj any) {
+	flw.setInclusion(obj, false)
+}
+
+func (flw *filteredListerWatcher) setInclusion(obj any, include bool) {
+	ws := obj.(*tenancyv1a1.Workspace)
+	mbwsName := ws.Name
+	mbwsNameParts := strings.Split(mbwsName, placement.WSNameSep)
+	if len(mbwsNameParts) != 2 {
```

Comment:
Is that enough condition for mb WS? Don't we need to compare the name to the actual MB name pattern ?